### ISSUE TITLE
Conda Support for 1.9.6 and 2.0.0.dev10

### ISSUE THREAD

**Issue by MarioJayakumar** [pygame/pygame#1953]
It would be helpful to have Conda support for pygame, specifically for versions 1.9.6 and 2.0.0.dev10. The version of pygame currently on Conda's CogSci channel is only 1.9.2. Having Conda support would be useful for users who prefer conda installations as opposed to pip, and for packages that depend on pygame.

**Comment by justinkterry** [comment]
We'd be willing to help do the packaging for this if needed. Ideally this would be an official thing, not random community repos like it is now.

**Comment by illume** [comment]
Why can't conda users use the pypi packages?

**Comment by justinkterry** [comment]
Generally, PyPI will install to a different Python environment than Python will, causing all sorts of problems.

In our case, we maintain a machine learning package that includes games made in PyGame. ML is one of the main communities that uses Conda, so we want to release on their, and we can't without a more current version of PyGame being on Conda.

**Comment by illume** [comment]
From my reading(but not actually using) it's valid and supported to use wheel files in conda.

Here's an example of how to install pygame into a conda environment:
```python
conda create --name pygame python=3.8
conda activate pygame
python3 -m pip install --upgrade pygame==2.0.0.dev10
python3 -m pygame.examples.aliens
```

Is it possible to depend on a pypi package from a conda package?

**Comment by MarioJayakumar** [comment]
Currently there is no way to specify a pypi dependency for a conda package, as all dependencies must be hosted on a conda channel.

**Comment by illume** [comment]
I don't want to support conda specific packages of pygame, but if someone does I'd suggest considering trying to reuse the existing packages. Because doing all that for num_python_versions * num_platforms is a lot of tedious work.

I'll leave this issue open for a few days to see if any other pygame developers are interested in helping out.

**Comment by justinkterry** [comment]
I don't think I understand your comment about trying to reuse existing packages?

Also we've done a bunch of work with Conda releases and would be willing to work on officially supporting PyGame on Conda.

**Comment by MyreMylar** [comment]
Since you seem like an expert.

Speaking for myself, without researching it I have no real idea what the current purpose of conda is so I don't know how to evaulate the scale of the task you are talking about. Is it another python packaging solution? Some other generic packaging solution? What does it want from other projects to support it? If it's a python thing (as the name suggests ) why hasn't Conda already adapted itself to using existing popular build processes like wheels and PyPI? Who uses it instead of using PyPI? Why does it exist? What other python projects on GitHub support conda in the way you are envisaging? - it would be useful to see what they are doing.

Speaking aesthetically/maintenance wise, I'd rather not have more config files in the repo in general - particularly as we just got through figuring out how to condense a bunch of them down into `setup.cfg`. It doesn't seem like conda supports setup.cfg, but if it does I'd prefer that to yet another config file. I'm already annoyed enough with black wanting it's own file to not support it.

Maintenance wise - what does this process involve? We have a lot of issues with pygame already and if there are going to be a pile of difficulties getting it to work with conda in the first place and then keeping it working with conda, that would be useful information. As illume mentions there are a lot of dependencies for pygame, and they tend to break fairly regularly on different combinations of platforms, python versions, SDL versions and so on. Adding conda versions on top of that sounds like a bag of not much fun, if we don't have someone who wants to regularly maintain it into the future (if that is what it takes).

How does this existing version of pygame in conda work, how did it get to wherever it is? If the conda process is annoying, would it not just be simpler to wait for the release of pygame version 2.0.0 and then stick that into conda as a one-time upgrade to replace the really old version?

**Comment by cruxicheiros** [comment]
Hey, I'm actually using conda for development for the latest version of pygame right now. I set up successfully today. Here's how I did it, for future reference:

- Made a new conda environment
- Installed the required dependencies from apt for building and running pygame
- Ran `conda install pip` which allows me to install packages from pip into conda rather than using conda's package manager - then later used `pip` to install other required python packages like numpy
- `python setup.py build`
- `pip install .` in the directory where setup.py is.

That's it. It was actually surprisingly painless in that aspect. It's just another kind of virtual environment as far as I'm aware, unless you also want to distribute pygame on conda I don't know why it needs special measures?

**Comment by illume** [comment]
So... I looked up some of the dependencies:
- https://anaconda.org/conda-forge/sdl2
- https://anaconda.org/conda-forge/sdl2_mixer
- https://anaconda.org/conda-forge/sdl2_ttf
- https://anaconda.org/conda-forge/sdl2_image

If those are working ok, and well maintained then I guess it would be fairly trivial to package pygame for conda. I didn't look up all the dependencies, but that's a pretty good start :)

**Comment by justinkterry** [comment]
@MyreMylar This is fairly easily Googleable, but I'll explain here:

Conda is an alternative package manager to pip. It doesn't use wheels, but it does still use setup.py. It has advantages over pip by natively supporting virtual environments, allowing for more easily portable installations across clusters, and lots of super useful things for BLAS and CUDA. Conda was actually created by the creator of NumPy, and is the default package manager for most workloads in scientific computing, data science and machine learning, and is most more useful for those workloads for all the reasons above. PyTorch (one of the two major deep learning libraries) explicitly supports conda before pip, more tutorials in the space use conda than pip, and scientific IDEs like Spyder only integrate with conda and not pip.

The main driving force behind putting PyGame on conda, and the reason it's a thing for us, is that learning to play games with machine learning is a very big thing, and machine learning people primarily use conda (for legitimate reasons).

The main thing conda would add is another place to release builds to, in addition PyPI/pip. It shouldn't require any major changes.

**Comment by robertpfeiffer** [comment]
> The main driving force behind putting PyGame on conda, and the reason it's a thing for us, is that learning to play games with machine learning is a very big thing, and machine learning people primarily use conda (for legitimate reasons).

The implications are obvious: PyGame contributors should publish a paper that doubles as a tutorial on PyGame as a tool/environment for machine learning, and link that on the home page.

**Comment by MyreMylar** [comment]
Thanks for the rundown @justinkterry and @cruxicheiros sounds like conda is embedded firmly in the data scientist space who are likely less comfortable with pypi and pip, so there is a reason to try and get pygame out to that group.

I had a look at NumPy's [GitHub repo](https://github.com/numpy/numpy) since it sounds like they are one of the main drivers of conda but I can't find this `meta.yaml` file mentioned earlier, searching for `conda` in there was not immediately illuminating either. Though it looks like numpy has forked it's own version of the old distutils library (something that pygame also still makes use of) which may provide some clues. 

I also had a peek at numpy's pypi page and, what I assume is the corresponding page on the anaconda cloud:

https://anaconda.org/anaconda/numpy
https://pypi.org/project/numpy/

It looks like the anaconda cloud is a version behind pypi? I guess this means that numpy also doesn't have an automated process to upload new versions which makes me curious since I suspect numpy has a more similar set of distribution problems to pygame than the average python library.

**Comment by MarioJayakumar** [comment]
So numpy uploads to the conda-forge channel. Conda-forge is basically a large repository that serves almost like a default channel in Conda. You can find a recipe here: https://github.com/conda-forge/numpy-feedstock/tree/master/recipe

However, I am not sure how they use their CI to upload to conda.

**Comment by MyreMylar** [comment]
Ah, so there is a separate GitHub repository that handles building and uploading projects to the conda cloud. 

So the best solution would be adding a `pygame-feedstock` repo on there? Looks  like they already have tons of them for different projects, the website implies that the feedstocks are all auto uploaded to the conda cloud from there:

https://conda-forge.org/

This is the process to create one it seems:

https://conda-forge.org/#add_recipe

Does that sound right conda experts?

**Comment by MarioJayakumar** [comment]
Yes, that is the process of adding pygame to the conda-forge channel. Additionally if you would instead like to upload it to your own channel you can do so by creating an account on conda cloud and uploading it there.

**Comment by justinkterry** [comment]
Any progress on this issue?

**Comment by MyreMylar** [comment]
This would be a good thing to do post the 2.0 release to get 2.0 onto conda.

**Comment by MyreMylar** [comment]
I'm really the worst person to coordinate with on this topic 🙂 on top of knowing nothing about conda I haven't had much time for pygame in the last 8 months and have not been involved with the packaging process at all when I had more time. @illume was usually the package master in the releases I was most heavily contributing to. I think @Starbuck5 has been been doing more stuff recently too but I am a bit out of the loop.

It seems like something that should happen, but as you say needs someone with passion for conda and time who works on pygame too.

**Comment by jkterry1** [comment]
@MyreMylar @illume @Starbuck5 Could we please coordinate to get pygame on Conda? I have a bunch of experience packaging python libraries with 10s of millions of downloads (e.g. I'm the maintainer of Gym https://github.com/openai/gym) and can help and @rohan138 has volunteered to help get this done in create detail. This is becoming increasingly important for another somewhat large library I'm in charge of (https://github.com/PettingZoo-Team/PettingZoo). I'd really like to get this done and me and Rohan can provide whatever help is needed to this end.

**Comment by robertpfeiffer** [comment]
I am looking through the docs and I don't understand how conda handles ISA?

**Comment by Starbuck5** [comment]
ISA means architecture? ARM vs x86_64?

@jkterry1, this is obviously something you want to happen, and have wanted to happen for a while, based on activity from this thread.

The thing is, pygame doesn’t have the resources of time and motivation needed to get more environments satisfied, especially right now. We’re having enough trouble getting wheels that work on Mac for 2.0.2 right now, and universal2 wheels seem incredibly daunting.

However, you seem to have the know how and motivation for it, unlike us, so I think you should start it up as a third party pygame distribution, with our blessing. Several pygame contributors are active on the pygame community discord, so you can to us more conversationally there, if you have questions. It’s at discord.gg/pygame

**Comment by robertpfeiffer** [comment]
> ISA means architecture? ARM vs x86_64?

Yeah, there is something about linux-64 and osx-64, but how does it handle ARM vs x86? I mean theoretically even windows runs on ARM. I read the docs and I can't make sense of them. Maybe I don't understand what kind of person the conda docs were written for, or I am reading the wrong docs...

**Comment by jkterry1** [comment]
I still plan to do Conda support eventually for Gym and PettingZoo, but this has been postponed for awhile until I put everything in that ecosystem on Conda.

Conda docs were written for data scientists to answer a question from awhile ago by the way.
